### ISSUE TITLE
Mention predefined keyboard shortcuts in the docs on event-handling

### ISSUE THREAD

**Issue by cacti77** [matplotlib/matplotlib#11472]
I've just wasted a whole day trying to debug my event-handling code that intercepted the 'key_press_event'.  In the end it turned out the key I was trying to handle ('l') was a known keyboard shortcut for Matplotlib at https://matplotlib.org/users/navigation_toolbar.html#navigation-keyboard-shortcuts and that all I had to do was use a different letter instead: a one-character fix!

I might have been less likely to have made this mistake had there been some reference to the pre-defined keyboard shortcuts somewhere in the event-handling tutorial at https://matplotlib.org/users/event_handling.html. Perhaps have a note next to the entries for the 'key_press_event' and 'key_release_event' in the table of events under https://matplotlib.org/users/event_handling.html#event-connections ? In general, though, the Matplotlib docs are superb.

**Comment by jklymak** [comment]
So the predefined shortcuts take precedence over user-defined?  That seems a poor API choice on our part - I would think we should check user defined first and then the defaults.

**Comment by jklymak** [comment]
Can you make a short self-contained example that has the problem?

**Comment by ImportanceOfBeingErnest** [comment]
This seems similar but not as bad as https://github.com/matplotlib/matplotlib/issues/11432, because at least the <kbd>l</kbd> key is documented as a built-in short-cut. Maybe just not visibly enough. I would support adding such note in the event handling guide.  

@cacti77 Are you willing to make a suggestion on this and submit it as a Pull-Request?

> So the predefined shortcuts take precedence over user-defined?

No, they are both registered indendently. So if you use the <kbd>l</kbd> key for some custom event, the built-in log scale toggle **and** the user defined event will happen both.

**Comment by jklymak** [comment]
^ Thanks; but that still seems incorrect to me.  If the user sets "l" to mean something, thats what it should mean, and our old keymap should be ignored.

**Comment by ImportanceOfBeingErnest** [comment]
OK, but how? The "keymap" is just some callback with a lot of `if event.key == "a":` and similar for every other letter. So there is no way of knowing which letter will do what at the end or if some key exists in some other callback as well.

**Comment by cacti77** [comment]
@jklymak Here's a short example:

	import matplotlib.pyplot as plt
	import networkx as nx
        
	def draw_network():
		plt.cla()
		plt.axis('off')
		nx.draw_networkx_nodes(G, pos, node_size=300, node_color=(0, 0, 0.9), edgecolors='black')
		nx.draw_networkx_labels(G, pos, font_color='red', font_family='arial', font_size=10)
		nx.draw_networkx_edges(G, pos, node_size=300, arrowstyle='->', arrowsize=15, edge_color='black', width=1)

	def on_key_press(event):
		global pos
		pos = nx.layout.spring_layout(G)
		draw_network()
		plt.draw()

	G = nx.DiGraph()
	G.add_nodes_from(['A', 'B', 'C', 'D', 'E'])
	G.add_edges_from([('A','B'), ('B','C'), ('B', 'D'), ('D', 'E')])

	pos = nx.layout.spring_layout(G, random_state=779)
	draw_network()

	canvas = plt.gcf().canvas
	canvas.mpl_connect('key_press_event', on_key_press)
	plt.show()

Run it. Zoom in on a portion of the network using the magnifying glass icon in the toolbar. Then press 'h'. Not only is the network re-laid out as intended but the view is reset to the home screen because of the predefined shortcut for 'h'.

My original code used the key 'l' to re-layout the network. Unfortunately, in stripping it down to this example, I lost something crucial that causes the y-axis to toggle between linear and log scales as per the Matplotlib shortcut. Annoyingly, I can't reproduce that effect with this example. It was really hard to debug because I didn't know about the predefined keyboard shortcuts; I'd hidden the axes so it wasn't obvious a linear-log conversion was happening; plus the predefined keyboard shortcut for 'l' only takes effect if the mouse happens to be over the axes.

I'm using Matplotlib 2.2.2, Python 2.7.13 and Windows 7.

**Comment by cacti77** [comment]
I agree with @jklymak that user-defined key events should override predefined ones. If both the predefined and user-defined callbacks are executed really unexpected things can happen, as I found out! Also, if both events happen, what order do they occur in?

Finally I want to choose the letters for my keyboard shortcuts. Matplotlib has consumed a large portion of the alphabet already for its own shortcuts, leaving only a few ones left for me to use. For example, I wanted to use 'h' to display help information, and 'l' to re-layout the network. Now I'm forced to use other letters which seem less intuitive to me. This problem will worsten whenever Matplotlib decides to add more keyboard shortcuts in future, plus break developers' previously working user-defined shortcut behaviour.

On the other hand some apps may already know about and use this combined pre-defined + user-defined effect and so changing this might break those apps.

If you're going to keep the current behaviour it would be good if there was some way of disabling or overriding the predefined shortcuts easily, and clearly document how to do so in the event-handling tutorial.

I don't think the docs should be updated until the intended behaviour has been agreed.

Re PRs I don't really know how to do those yet and I'm short of time. But I don't mind commented on changes to the docs.

**Comment by ImportanceOfBeingErnest** [comment]
I think the crucial part is really about knowing vs. not knowing how to do things. 

Once you know about predefined short-cuts, it's easy to handle them. In this case it seems you want to use <kbd>l</kbd> and <kbd>h</kbd> for your own purposes. The first step would be to either disable the predefined functionality or map it to some other key, e.g.

```
plt.rcParams["keymap.yscale"] = ""
plt.rcParams["keymap.home"] = "r", "home"
```

Next you can define <kbd>l</kbd> and <kbd>h</kbd> to do what you like them to do in a callback. 

<details>
 <summary>Some complete example</summary>

```
import matplotlib.pyplot as plt
plt.rcParams["keymap.yscale"] = ""
plt.rcParams["keymap.home"] = "r", "home"

fig, ax = plt.subplots()

def callback(evt):
    if evt.key == "l":
        print("l pressed")
    if evt.key == "h":
        print("h pressed")

fig.canvas.mpl_connect("key_press_event", callback)
plt.show()
```

</details>

<hr>

> Also, if both events happen, what order do they occur in?

That is not defined. They are stored in a dictionary, so whichever callback comes first when looping over that dictionary is executed first. 

I think the behaviour as it is is fine (except for the issue in #11432), but should indeed be better documented, to allow users to directly see the option of (re)defining keys.

**Comment by cacti77** [comment]
Thanks for the simple example; that's really useful!  I can live with the current behaviour as long as it's well documented and it's easy to override, as you have shown in your example. Perhaps that example should go in the event-handling docs.  Maybe even put it in a new short section specifically on the handling of keyboard events.

**Comment by asvela** [comment]
It seems this was never added to the documentation? I spent a few hours wondering why `l` was making my axes swap between log and linear in addition to the other changes to the y axis that I was trying to do..

(Then I also discovered that the number keys do things too! I thought the number keys should be safe at least as they are not in the rcParams, but [deep in backend_bases.py](https://github.com/matplotlib/matplotlib/blob/fc18cccbb9122ad6d7fc6156ec8f6c455e408d8c/lib/matplotlib/backend_bases.py#L2618) I finally understood what was going on. Good these are being removed from v3.3.)
